internal/cmd/local/k8s/load_images.go

```diff
@@ -0,0 +1,208 @@
+package k8s
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+
+	"github.com/airbytehq/abctl/internal/cmd/local/docker"
+	"github.com/airbytehq/abctl/internal/common"
+	"github.com/docker/docker/api/types/filters"
+	"github.com/docker/docker/api/types/image"
+	"github.com/pterm/pterm"
+	nodeslib "sigs.k8s.io/kind/pkg/cluster/nodes"
+	"sigs.k8s.io/kind/pkg/cluster/nodeutils"
+	"sigs.k8s.io/kind/pkg/exec"
+	"sigs.k8s.io/kind/pkg/fs"
+)
+
+// loadImages pulls and loads images into the kind cluster.
+// It will pull all images in parallel, skip any images that already exist on the nodes,
+// save the rest to an image archive (tar file), and load archive onto the nodes.
+func loadImages(ctx context.Context, dockerClient docker.Client, nodes []nodeslib.Node, images []string) error {
+
+	// Pull all the images via "docker pull", in parallel.
+	var wg sync.WaitGroup
+	wg.Add(len(images))
+	for _, img := range images {
+		pterm.Info.Printfln("Pulling image %s", img)
+
+		go func(img string) {
+			defer wg.Done()
+			r, err := dockerClient.ImagePull(ctx, img, image.PullOptions{})
+			if err != nil {
+				pterm.Debug.Printfln("error pulling image %s", err)
+				// image pull errors are intentionally dropped because we're in a goroutine,
+				// and because we don't want to interrupt other image pulls.
+			}
+			defer r.Close()
+			io.Copy(io.Discard, r)
+		}(img)
+	}
+	wg.Wait()
+
+	// The context could be canceled by now. If so, return early.
+	if ctx.Err() != nil {
+		return ctx.Err()
+	}
+
+	// Determine which images need to be loaded onto the nodes.
+	needed := determineImagesForLoading(ctx, dockerClient, images, nodes)
+	if len(needed) == 0 {
+		return nil
+	}
+
+	// for _, node := range nodes {
+	// 	for _, img := range needed {
+
+	// 		pterm.Debug.Printfln("saving image %s into node %s", img, node)
+
+	// 		r, err := dockerClient.ImageSave(ctx, images)
+	// 		if err != nil {
+	// 			pterm.Debug.Printfln("error loading image %s into node %s: %s", img, node, err)
+	// 			continue
+	// 		}
+
+	// 		pterm.Debug.Printfln("loading image %s into node %s", img, node)
+
+	// 		err = nodeutils.LoadImageArchive(node, r)
+	// 		if err != nil {
+	// 			pterm.Debug.Printfln("error loading image %s into node %s: %s", img, node, err)
+	// 		}
+	// 	}
+	// }
+
+	// Save all the images to an archive, images.tar
+	imagesTarPath, err := saveImageArchive(ctx, dockerClient, needed)
+	if err != nil {
+		return fmt.Errorf("failed to save image archive: %w", err)
+	}
+	// defer os.RemoveAll(imagesTarPath)
```

Comment:
I think we want to uncomment this, as we should clean up any files we create.

Reply:
oops!